compiler/src/org.graalvm.compiler.replacements.amd64/src/org/graalvm/compiler/replacements/amd64/AMD64StringLatin1InflateNode.java

```diff
@@ -0,0 +1,106 @@
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.graalvm.compiler.replacements.amd64;
+
+import static org.graalvm.compiler.nodeinfo.InputType.Memory;
+import static org.graalvm.compiler.nodeinfo.NodeSize.SIZE_512;
+import static org.graalvm.compiler.nodeinfo.NodeCycles.CYCLES_UNKNOWN;
+
+import org.graalvm.compiler.core.common.type.StampFactory;
+import org.graalvm.compiler.graph.NodeClass;
+import org.graalvm.compiler.graph.Node.OptionalInput;
+import org.graalvm.compiler.lir.gen.LIRGeneratorTool;
+import org.graalvm.compiler.nodeinfo.NodeInfo;
+import org.graalvm.compiler.nodes.FixedWithNextNode;
+import org.graalvm.compiler.nodes.NamedLocationIdentity;
+import org.graalvm.compiler.nodes.ValueNode;
+import org.graalvm.compiler.nodes.ValueNodeUtil;
+import org.graalvm.compiler.nodes.memory.MemoryAccess;
+import org.graalvm.compiler.nodes.memory.MemoryCheckpoint;
+import org.graalvm.compiler.nodes.memory.MemoryNode;
+import org.graalvm.compiler.nodes.spi.LIRLowerable;
+import org.graalvm.compiler.nodes.spi.NodeLIRBuilderTool;
+import org.graalvm.word.LocationIdentity;
+import org.graalvm.word.Pointer;
+
+import jdk.vm.ci.meta.JavaKind;
+
+
+@NodeInfo(allowedUsageTypes = Memory, size = SIZE_512, cycles = CYCLES_UNKNOWN)
+
+public final class AMD64StringLatin1InflateNode
+        extends FixedWithNextNode implements LIRLowerable, MemoryCheckpoint.Multi, MemoryAccess {
+
+    public static final NodeClass<AMD64StringLatin1InflateNode> TYPE = NodeClass.create(AMD64StringLatin1InflateNode.class);
+
+    @Input private ValueNode src;
+    @Input private ValueNode dst;
+    @Input private ValueNode len;
+
+    @OptionalInput(Memory) private MemoryNode lla; // Last access location registered.
+
+    // "java.lang.StringLatin1.inflate([BI[CII)V"
+
+    // public static void inflate(byte[] src, int src_indx, char[] dst, int dst_indx, int len)
+
+    public AMD64StringLatin1InflateNode(ValueNode src, ValueNode dst, ValueNode len) {
+        super(TYPE, StampFactory.forVoid());
+        this.src = src;
+        this.dst = dst;
+        this.len = len;
+    }
+
+    @Override
+    public LocationIdentity getLocationIdentity() {
+        // Model read access via 'src' using:
+        return NamedLocationIdentity.getArrayLocation(JavaKind.Byte);
+    }
+
+    @Override
```

Comment:
You can implement `MemoryCheckpoint.Single` instead and then you won't need this override.

Reply:
Hmm, I actually did start off trying to use "Single" but could not make it produce the proper graph... perhaps I didn't get the annotation/properties right to begin with...

Reply:
@tkrodriguez I think the issue is that `AMD64StringLatin1InflateNode` reads from a `byte[]` and writes to a `char[]`. `getLocationIdentity` is modelling the read (by implementing `MemoryAccess`) and `getLocationIdentities()` is modelling the write (by implementing `MemoryCheckpoint.Multi`). It's an interesting trick to disambiguate modelling the read and write in the memory graph.

Reply:
Yes, but perhaps not necessary?

Reply:
I think the current code with `Multi` is correct. While we *might* get lucky and not break anything if we use `Single` because this node is fixed, i think it's better to model things precisely so that the rest of the compiler can make correct decisions.

Reply:
Yes I missed that it was reading and writing different types.  I think a comment explaining this slightly odd usage would be a good idea.